api/napi/src/types/model.rs

```diff
@@ -1,36 +1,177 @@
 // Copyright © SixtyFPS GmbH <info@slint.dev>
 // SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
 
-use i_slint_core::model::{Model, ModelRc, VecModel};
-use napi::bindgen_prelude::External;
-use napi_derive::napi;
+use std::rc::{Rc, Weak};
+
+use i_slint_compiler::langtype::Type;
+use i_slint_core::model::Model;
+use napi::{bindgen_prelude::Object, Env, JsFunction, JsNumber, JsUnknown, NapiRaw, Result};
 use slint_interpreter::Value;
 
-#[napi(js_name = Model)]
+use crate::{to_js_unknown, to_value, RefCountedReference};
+
 pub struct JsModel {
-    inner: ModelRc<Value>,
+    model: RefCountedReference,
+    env: Env,
+    notify: i_slint_core::model::ModelNotify,
+    data_type: Type,
 }
 
-impl From<ModelRc<Value>> for JsModel {
-    fn from(model: ModelRc<Value>) -> Self {
-        Self { inner: model }
+impl JsModel {
+    pub fn new<T: NapiRaw>(env: Env, model: T, data_type: Type) -> napi::Result<Rc<Self>> {
+        let js_model = Rc::new(Self {
+            notify: Default::default(),
+            env,
+            model: RefCountedReference::new(&env, model)?,
+            data_type,
+        });
+
+        let notify = JsSlintModelNotify { model: Rc::downgrade(&js_model) };
+
+        js_model.model.get::<Object>()?.set("notify", notify)?;
+
+        Ok(js_model)
+    }
+
+    pub fn model(&self) -> &RefCountedReference {
+        &self.model
+    }
+}
+
+impl Model for JsModel {
+    type Data = slint_interpreter::Value;
+
+    fn row_count(&self) -> usize {
+        let model: Object = self.model.get().unwrap();
+        model
+            .get::<&str, JsFunction>("rowCount")
+            .ok()
+            .and_then(|callback| {
+                callback.and_then(|callback| callback.call::<JsUnknown>(Some(&model), &[]).ok())
+            })
+            .and_then(|res| res.coerce_to_number().ok())
+            .map(|num| num.get_uint32().ok().map_or(0, |count| count as usize))
+            .unwrap_or_default()
+    }
+
+    fn row_data(&self, row: usize) -> Option<Self::Data> {
+        let model: Object = self.model.get().unwrap();
+        model
+            .get::<&str, JsFunction>("rowData")
+            .ok()
+            .and_then(|callback| {
+                callback.and_then(|callback| {
+                    callback
+                        .call::<JsNumber>(
+                            Some(&model),
+                            &[self.env.create_double(row as f64).unwrap()],
+                        )
+                        .ok()
+                })
+            })
+            .and_then(|res| to_value(&self.env, res, self.data_type.clone()).ok())
```

Comment:
When using a `ModelIterator` in Rust on a `Model` implementation, this is the `next()` implementation that gets called:

```rust
    fn next(&mut self) -> Option<Self::Item> {
        let row = self.row;
        if self.row < self.model.row_count() {
            self.row += 1;
        }
        self.model.row_data(row)
    }
```

This is where the run-time library relies on the ability to call `row_data()` with an out-of-bounds row and expects to get `None` back, which will also end the iteration.

The interpreter, when calling `set_property` will perform some type compatibility checks, and for that it creates an iterator on a model and iterates through all rows. That's why the test ends up calling this with "out of bounds".

I suggest to check if `res.get_value()` matches `Ok(napi::ValueType::Undefined)` and if it does, then return `None` for `row_data`.